csharp/src/Ice/InputStream.cs

```diff
@@ -2104,6 +2115,23 @@ private string ReadTypeId(bool isIndex)
             return cls;
         }
 
+        private Type? ResolveClass(int compactId)
+        {
+            Type? cls = null;
+            if (_compactIdCache?.TryGetValue(compactId, out cls) != true)
```

Comment:
!TryGetValue doesn't work here because the null coalescing, but it will work if we move lazy initialization up

```
Type? cls;
_typeIdCache ??= new Dictionary<string, Type?>(); // Lazy initialization
if (!_typeIdCache.TryGetValue(typeId, out cls))
{
    // Not found in typeIdCache
    try
    {
        cls = _classResolver?.Invoke(typeId);
        _typeIdCache.Add(typeId, cls);
    }
    catch (Exception ex)
    {
        throw new NoClassFactoryException("no class factory", typeId, ex);
    }
}
return cls;
```

Reply:
"<nullable> != true" is correct here, but I guess not readable enough?
the simpler and possibly more readable alternative is
```
if (_compactIdCache == null || !_compactIdCache.TryGetValue(...))
{
    // not found
}
```
"<nullable> == false" does not have the semantics as it returns "false" when nullable is null.

In this particular situation, we can probably lazy initialize earlier as Jose suggested to make the core more readable.

Reply:
Fixed with
```
if (_compactIdCache == null || !_compactIdCache.TryGetValue(...))
```

Reply:
I think moving `_compactIdCache` lazy initialization is a bit clear, it avoid to check null twice one in the if condition and second in the lazy initialization

Reply:
The slight downside with moving the lazy initialization first is that we then search the just created map even though it's empty. In any case, this code will be later refactored out of InputStream into Communicator.